lte/gateway/python/magma/kernsnoopd/ebpf/byte_count.bpf.c

```diff
@@ -23,28 +24,48 @@ BPF_HASH(dest_counters, struct key_t);
 
 // Attach kprobe for the `tcp_sendmsg` syscall
 int kprobe__tcp_sendmsg(struct pt_regs *ctx, struct sock *sk, struct msghdr *msg, size_t size) {
-  u16 dport = 0, family = sk->__sk_common.skc_family;
+  u16 family = sk->sk_family;
+
+  // both IPv4 and IPv6
+  if (family == AF_INET || family == AF_INET6) {
+    struct key_t key = {};
 
-  // only IPv4
-  if (family == AF_INET) {
-    struct key_t key;
-    key.daddr = sk->__sk_common.skc_daddr;
+    // read destination IP address and port
+    if (family == AF_INET) {
+      key.daddr = sk->sk_daddr;
+    }
+    else if (family == AF_INET6) {
+      bpf_probe_read(&key.daddr, sizeof(key.daddr), &sk->sk_v6_daddr.s6_addr32);
+    }
+    u16 dport = sk->sk_dport;
+    key.dport = ntohs(dport);
+
+    // ignore packets destined for localhost unless dport is PROXY_PORT
 
-    // ignore packets destined for localhost
     // IPv4 127.0.0.1 == 16777343, 0.0.0.0 == 0
-    if ((key.daddr == 16777343 || key.daddr == 0)) {
-      return 0;
+    if (family == AF_INET && key.dport != {{PROXY_PORT}}) {
+      if (key.daddr == 0x100007F || key.daddr == 0) {
+        return 0;
+      }
+    }
+    // IPv6 localhost has embedded IPv4 in case of control_proxy
+    // uint128 is split into two u64 and high is FFFF:"127.0.0.1"
+    else if (family == AF_INET6 && key.dport != {{PROXY_PORT}}) {
+      uint64_t low = (uint64_t)key.daddr;
+      uint64_t high = (key.daddr >> 64);
+      if (low == 0 && high == 0x100007FFFFF0000) {
+        return 0;
+      }
     }
 
-    // read binary name, pid, source and destination IP address and port
+    // read binary name, pid, and source port
     bpf_get_current_comm(key.comm, TASK_COMM_LEN);
```

Comment:
I missed it in last round, can we avoid reading the command name, since we can map the pid to service name?

Reply:
This comm is only used as backup when the captured PID is no longer valid (process exited). We also want to record arbitrary Linux binaries generating traffic. PID to service/binary name mapping will happen every `collect_interval` seconds. The minimum `collect_interval` right now is 10 seconds and can be configured to higher values. If we don't capture the comm here, we will miss short-lived, perhaps frequent processes generating significant traffic.